File: native-tests/android/app/src/main/java/com/cactus/nativetest/TestRunner.kt
package com.cactus.nativetest

import android.graphics.Color
import android.os.Handler
import android.os.Looper
import com.cactus.Cactus
import com.cactus.CompletionOptions
import com.cactus.Message
import java.io.File
import kotlin.math.sqrt

class TestRunner(
    private val onLog: (String, Int) -> Unit,
    private val onDone: () -> Unit
) {
    private val mainHandler = Handler(Looper.getMainLooper())

    fun run(modelPath: String, transcribeModelPath: String, assetsPath: String) {
        Thread { runTests(modelPath, transcribeModelPath, assetsPath) }.start()
    }

    private fun runTests(modelPath: String, transcribeModelPath: String, assetsPath: String) {
        log("=== Cactus Kotlin Wrapper Test ===")
        log("Model:            ${if (modelPath.isEmpty()) "(none - pass MODEL_PATH intent extra)" else modelPath}")
        log("Transcribe model: ${if (transcribeModelPath.isEmpty()) "(none)" else transcribeModelPath}")
        log("Assets:           ${if (assetsPath.isEmpty()) "(none)" else assetsPath}")

        if (modelPath.isEmpty()) {
            log("[FAIL] No model path provided")
            done(); return
        }

        // ---------------------------------------------------------------
        log("\n--- Test 1: Init ---")
        val model = try {
            Cactus.create(modelPath).also { log("[PASS] Model initialized") }
        } catch (e: Exception) {
            log("[FAIL] $e"); done(); return
        }

        // ---------------------------------------------------------------
        log("\n--- Test 2: Basic Completion ---")
        try {
            val result = model.complete(
                listOf(Message.user("Say hello in exactly 3 words.")),
                CompletionOptions(maxTokens = 20)
            )
            val text = result.text.trim()
            log("[PASS] \"$text\"")
            log("       prompt=${result.promptTokens} completion=${result.completionTokens} decode=${"%.1f".format(result.decodeTokensPerSecond)} tok/s")
        } catch (e: Exception) {
            log("[FAIL] $e")
        }

        // ---------------------------------------------------------------
        log("\n--- Test 3: Streaming Completion ---")
        try {
            var tokenCount = 0
            val result = model.complete(
                listOf(Message.user("Count from 1 to 5.")),
                CompletionOptions(maxTokens = 40),
                callback = { _, _ -> tokenCount++ }
            )
            if (tokenCount > 0) {
                log("[PASS] Streamed $tokenCount tokens: \"${result.text.trim()}\"")
            } else {
                log("[FAIL] onToken callback never fired")
            }
        } catch (e: Exception) {
            log("[FAIL] $e")
        }

        // ---------------------------------------------------------------
        log("\n--- Test 4: Tool Calling ---")
        try {
            val tools = listOf(mapOf(
                "name" to "get_weather",
                "description" to "Get the current weather for a location",
                "parameters" to mapOf(
                    "type" to "object",
                    "properties" to mapOf(
                        "location" to mapOf("type" to "string", "description" to "City name")
                    ),
                    "required" to listOf("location")
                )
            ))
            val result = model.complete(
                listOf(Message.user("What is the weather in Paris? Use the get_weather tool.")),
                CompletionOptions(maxTokens = 80),
                tools = tools
            )
            val calls = result.functionCalls
            if (!calls.isNullOrEmpty()) {
                val name = calls.first()["name"] as? String ?: "?"
                val args = calls.first()["arguments"] ?: emptyMap<String, Any>()
                log("[PASS] Tool call: $name($args)")
            } else {
                log("[SKIP] No tool call produced (model may not support function calling)")
            }
        } catch (e: Exception) {
            log("[FAIL] $e")
        }

        // ---------------------------------------------------------------
        log("\n--- Test 5: Tokenization ---")
        try {
            val tokens = model.tokenize("Hello, world!")
            if (tokens.isNotEmpty()) {
                log("[PASS] ${tokens.size} tokens")
            } else {
                log("[FAIL] Zero tokens returned")
            }
        } catch (e: Exception) {
            log("[FAIL] $e")
        }

        // ---------------------------------------------------------------
        log("\n--- Test 6: Embeddings ---")
        try {
            val embeddings = model.embed("Hello world")
            if (embeddings.isNotEmpty()) {
                val norm = sqrt(embeddings.fold(0f) { s, x -> s + x * x }).toDouble()
                log("[PASS] dim=${embeddings.size} norm=${"%.4f".format(norm)}")
            } else {
                log("[SKIP] Empty embeddings (model may not support)")
            }
        } catch (e: Exception) {
            log("[SKIP] Not supported by this model: $e")
        }

        // ---------------------------------------------------------------
        val audioPath = if (assetsPath.isEmpty()) "" else "$assetsPath/test.wav"
        val audioExists = audioPath.isNotEmpty() && File(audioPath).exists()

        if (transcribeModelPath.isNotEmpty()) {
            val transcribeModel = try {
                Cactus.create(transcribeModelPath)
            } catch (e: Exception) {
                log("\n--- Test 7: Transcription --- [FAIL] Could not init transcribe model: $e")
                log("--- Test 8: VAD --- [SKIP]")
                log("\n=== Done ===")
                model.close(); done(); return
            }

            log("\n--- Test 7: Transcription ---")
            if (audioExists) {
                try {
                    val isWhisper = transcribeModelPath.lowercase().contains("whisper")
                    val prompt = if (isWhisper) "<|startoftranscript|><|en|><|transcribe|><|notimestamps|>" else null
                    val result = transcribeModel.transcribe(audioPath, prompt = prompt)
                    log("[PASS] \"${result.text.trim()}\"")
                    log("       time=${"%.0f".format(result.totalTime)}ms")
                } catch (e: Exception) {
                    log("[FAIL] $e")
                }
            } else {
                log("[SKIP] test.wav not found at $audioPath")
            }

            log("\n--- Test 8: VAD ---")
            if (audioExists) {
                try {
                    val result = transcribeModel.vad(audioPath)
                    log("[PASS] ${result.segments.size} speech segment(s)")
                    for (seg in result.segments) {
                        log("       segment: ${seg.start}ms – ${seg.end}ms")
                    }
                } catch (e: Exception) {
                    log("[FAIL] $e")
                }
            } else {
                log("[SKIP] test.wav not found")
            }

            transcribeModel.close()
        } else {
            log("\n--- Test 7: Transcription --- [SKIP] (no TRANSCRIBE_PATH intent extra)")
            log("--- Test 8: VAD --- [SKIP] (no TRANSCRIBE_PATH intent extra)")
        }

        model.close()
        log("\n=== Done ===")
        done()
    }

    private fun log(message: String) {
        println(message)
        mainHandler.post { onLog(message, colorFor(message)) }
    }

    private fun done() {
        mainHandler.post { onDone() }
    }

    private fun colorFor(log: String) = when {
        log.startsWith("[PASS]") -> Color.parseColor("#4CAF50")
        log.startsWith("[FAIL]") -> Color.parseColor("#F44336")
        log.startsWith("[SKIP]") -> Color.parseColor("#FF9800")
        log.startsWith("===") -> Color.BLACK
        else -> Color.DKGRAY
    }
}

File: native-tests/ios/CactusSwiftTest/CactusSwiftTest/TestRunner.swift
import Foundation
import Combine

class TestRunner: ObservableObject {
    @Published var logs: [String] = []
    @Published var finished = false

    func run(modelPath: String, transcribeModelPath: String, assetsPath: String) {
        DispatchQueue.global(qos: .userInitiated).async {
            self.log("=== Cactus Swift Wrapper Test ===")
            self.log("Model:            \(modelPath.isEmpty ? "(none - set CACTUS_TEST_MODEL)" : modelPath)")
            self.log("Transcribe model: \(transcribeModelPath.isEmpty ? "(none)" : transcribeModelPath)")
            self.log("Assets:           \(assetsPath.isEmpty ? "(none)" : assetsPath)")

            guard !modelPath.isEmpty else {
                self.log("[FAIL] No model path provided")
                self.finish()
                return
            }

            // ---------------------------------------------------------------
            self.log("\n--- Test 1: Init ---")
            var model: Cactus?
            do {
                model = try Cactus(modelPath: modelPath)
                self.log("[PASS] Model initialized")
            } catch {
                self.log("[FAIL] \(error)")
                self.finish()
                return
            }

            // ---------------------------------------------------------------
            self.log("\n--- Test 2: Basic Completion ---")
            do {
                let result = try model!.complete(
                    messages: [.user("Say hello in exactly 3 words.")],
                    options: .init(maxTokens: 20)
                )
                let text = result.text.trimmingCharacters(in: .whitespacesAndNewlines)
                self.log("[PASS] \"\(text)\"")
                self.log("       prompt=\(result.promptTokens) completion=\(result.completionTokens) decode=\(String(format: "%.1f", result.decodeTokensPerSecond)) tok/s")
            } catch {
                self.log("[FAIL] \(error)")
            }

            // ---------------------------------------------------------------
            self.log("\n--- Test 3: Streaming Completion ---")
            do {
                var tokenCount = 0
                let result = try model!.complete(
                    messages: [.user("Count from 1 to 5.")],
                    options: .init(maxTokens: 40),
                    onToken: { _, _ in tokenCount += 1 }
                )
                if tokenCount > 0 {
                    self.log("[PASS] Streamed \(tokenCount) tokens: \"\(result.text.trimmingCharacters(in: .whitespacesAndNewlines))\"")
                } else {
                    self.log("[FAIL] onToken callback never fired")
                }
            } catch {
                self.log("[FAIL] \(error)")
            }

            // ---------------------------------------------------------------
            self.log("\n--- Test 4: Tool Calling ---")
            do {
                let tools: [[String: Any]] = [[
                    "name": "get_weather",
                    "description": "Get the current weather for a location",
                    "parameters": [
                        "type": "object",
                        "properties": [
                            "location": ["type": "string", "description": "City name"]
                        ],
                        "required": ["location"]
                    ] as [String: Any]
                ]]
                let result = try model!.complete(
                    messages: [.user("What is the weather in Paris? Use the get_weather tool.")],
                    options: .init(maxTokens: 80),
                    tools: tools
                )
                if let calls = result.functionCalls, !calls.isEmpty {
                    let name = calls.first?["name"] as? String ?? "?"
                    let args = calls.first?["arguments"] as? [String: Any] ?? [:]
                    self.log("[PASS] Tool call: \(name)(\(args))")
                } else {
                    self.log("[SKIP] No tool call produced (model may not support function calling)")
                }
            } catch {
                self.log("[FAIL] \(error)")
            }

            // ---------------------------------------------------------------
            self.log("\n--- Test 5: Tokenization ---")
            do {
                let tokens = try model!.tokenize("Hello, world!")
                if tokens.count > 0 {
                    self.log("[PASS] \(tokens.count) tokens")
                } else {
                    self.log("[FAIL] Zero tokens returned")
                }
            } catch {
                self.log("[FAIL] \(error)")
            }

            // ---------------------------------------------------------------
            self.log("\n--- Test 6: Embeddings ---")
            do {
                let embeddings = try model!.embed(text: "Hello world")
                if embeddings.count > 0 {
                    self.log("[PASS] dim=\(embeddings.count) norm=\(String(format: "%.4f", Self.norm(embeddings)))")
                } else {
                    self.log("[SKIP] Empty embeddings (model may not support)")
                }
            } catch {
                self.log("[SKIP] Not supported by this model: \(error)")
            }

            // ---------------------------------------------------------------
            let audioPath = assetsPath.isEmpty ? "" : (assetsPath + "/test.wav")
            let audioExists = !audioPath.isEmpty && FileManager.default.fileExists(atPath: audioPath)

            if !transcribeModelPath.isEmpty {
                var transcribeModel: Cactus?
                do {
                    transcribeModel = try Cactus(modelPath: transcribeModelPath)
                } catch {
                    self.log("\n--- Test 7: Transcription --- [FAIL] Could not init transcribe model: \(error)")
                    self.log("--- Test 8: VAD --- [SKIP]")
                    self.log("\n=== Done ===")
                    self.finish()
                    return
                }

                self.log("\n--- Test 7: Transcription ---")
                if audioExists {
                    do {
                        let modelLower = transcribeModelPath.lowercased()
                        let prompt = modelLower.contains("whisper")
                            ? "<|startoftranscript|><|en|><|transcribe|><|notimestamps|>"
                            : nil
                        let result = try transcribeModel!.transcribe(audioPath: audioPath, prompt: prompt)
                        self.log("[PASS] \"\(result.text.trimmingCharacters(in: .whitespacesAndNewlines))\"")
                        self.log("       time=\(String(format: "%.0f", result.totalTime))ms")
                    } catch {
                        self.log("[FAIL] \(error)")
                    }
                } else {
                    self.log("[SKIP] test.wav not found at \(audioPath)")
                }

                self.log("\n--- Test 8: VAD ---")
                if audioExists {
                    do {
                        let result = try transcribeModel!.vad(audioPath: audioPath)
                        self.log("[PASS] \(result.segments.count) speech segment(s)")
                        for seg in result.segments {
                            self.log("       segment: \(seg.start)ms – \(seg.end)ms")
                        }
                    } catch {
                        self.log("[FAIL] \(error)")
                    }
                } else {
                    self.log("[SKIP] test.wav not found")
                }
            } else {
                self.log("\n--- Test 7: Transcription --- [SKIP] (no CACTUS_TEST_TRANSCRIBE_MODEL)")
                self.log("--- Test 8: VAD --- [SKIP] (no CACTUS_TEST_TRANSCRIBE_MODEL)")
            }

            self.log("\n=== Done ===")
            self.finish()
        }
    }

    private static func norm(_ v: [Float]) -> Float {
        v.reduce(0) { $0 + $1 * $1 }.squareRoot()
    }

    private func log(_ message: String) {
        print(message)
        DispatchQueue.main.async { self.logs.append(message) }
    }

    private func finish() {
        DispatchQueue.main.async {
            self.finished = true
            exit(0)
        }
    }
}
